File: FinbourneLRUCacheJava/src/com/finbourne/LRUCacheTest.java
package com.finbourne;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.LinkedList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class LRUCacheTest {
    LRUCache cache = LRUCache.getInstance();

    @BeforeEach
    void beforeEach() {
        cache.clear();
        cache.resize(99);
    }

    @Test
    void isSingleton() {
        assertSame(cache, LRUCache.getInstance());
    }

    @Test
    void basicPutAndGet() {
        cache.put("x", "xx");
        cache.put(true, "false");
        assertEquals(cache.size(), 2);
        assertEquals(cache.get("x"), "xx");
        assertEquals(cache.get(true), "false");
    }

    @Test
    void basicPutAndRemove() {
        cache.put("x", "xx");
        cache.put(true, "false");
        assertEquals(cache.size(), 2);
        assertEquals(cache.remove(true), "false");
        assertNull(cache.remove(99));
        assertEquals(cache.size(), 1);
        assertNull(cache.get(true));
    }

    @Test
    void maxCapacity() {
        cache.resize(1);
        cache.put(1, 11);
        cache.put(2, 22);
        assertEquals(cache.size(), 1);
        assertNull(cache.get(1));
        assertEquals(cache.get(2), 22);
    }

    @Test
    void getShouldRefreshLRU() {
        cache.put(1, 11);
        cache.put(2, 22);
        cache.get(1);
        cache.resize(1);
        assertEquals(cache.size(), 1);
        assertNull(cache.get(2));
        assertEquals(cache.get(1), 11);
    }

    @Test
    void illegalCapacity() {
        assertThrows(IllegalArgumentException.class, () -> cache.resize(-2));
    }

    @Test
    void callBackOnEviction() {
        final int[] x = {0};
        cache.put(1, 11, new EvictionCallback() {
            @Override
            public void apply(Object key, Object value) {
                x[0] = (int) value;
            }
        });
        cache.remove(1);
        assertEquals(x[0], 11);
    }

    @Test
    void resizeReturnsListOfEvicted() {
        cache.put(1, 11);
        cache.put(2, 22);
        cache.put(3, 33);
        List<Object> expected = new LinkedList<>();
        expected.add(11);
        expected.add(22);
        assertEquals(expected, cache.resize(1));
    }
}

File: FinbourneLRUCacheCSharp/LRUCacheTest/LRUCacheTest.cs
using LRUCache;

namespace LRUCacheTest
{
    [TestClass]
    public class LRUCacheTest
    {

        LRUCache.LRUCache cache = LRUCache.LRUCache.Instance;

        [TestInitialize]
        public void BeforeEach()
        {
            cache.Clear();
            cache.Resize(99);
        }

        [TestMethod]
        public void IsSingleton()
        {
            Assert.AreSame(cache, LRUCache.LRUCache.Instance);
        }

        [TestMethod]
        public void BasicPutAndGet()
        {
            cache.Put("x", "xx");
            cache.Put(true, false);
            Assert.AreEqual(cache.Size(), 2);
            Assert.AreEqual(cache.Get("x"), "xx");
            Assert.AreEqual(cache.Get(true), false);
        }

        [TestMethod]        
        public void BasicPutAndRemove()
        {
            cache.Put("x", "xx");
            cache.Put(true, "false");
            Assert.AreEqual(cache.Size(), 2);
            Assert.AreEqual(cache.Remove(true), "false");
            Assert.IsNull(cache.Remove(99));
            Assert.AreEqual(cache.Size(), 1);
            Assert.IsNull(cache.Get(true));
        }

        [TestMethod]
        public void MaxCapacity()
        {
            cache.Resize(1);
            cache.Put(1, 11);
            cache.Put(2, 22);
            Assert.AreEqual(cache.Size(), 1);
            Assert.IsNull(cache.Get(1));
            Assert.AreEqual(cache.Get(2), 22);
        }


        [TestMethod]
        public void GetShouldRefreshLRU() 
        {
            cache.Put(1, 11);
            cache.Put(2, 22);
            cache.Get(1);
            cache.Resize(1);
            Assert.AreEqual(cache.Size(), 1);
            Assert.IsNull(cache.Get(2));
            Assert.AreEqual(cache.Get(1), 11);
        }

        [TestMethod]
        public void IllegalCapacity()
        {
            Assert.ThrowsException<ArgumentException>(() => { cache.Resize(-1); });
        }

        [TestMethod]
        public void CallBackOnEviction()
        {
            int x = 0;
            cache.Put(1, 11, (key, value) =>
            {
                x = (int)value;
            });
            cache.Remove(1);
            Assert.AreEqual(x, 11);
        }

        [TestMethod]
        public void ResizeReturnsListOfEvicted()
        {
            cache.Put(1, 11);
            cache.Put(2, 22);
            cache.Put(3, 33);
            Dictionary<object, object> expected = new();
            expected.Add(1, 11);
            expected.Add(2, 22);
            Assert.IsTrue(expected.SequenceEqual(cache.Resize(1)));
        }
    }
}